src/main/scala/cromwell/util/docker/DockerHash.scala

```diff
@@ -0,0 +1,119 @@
+
+import cromwell.util.StringUtil._
+
+import scala.collection.generic.CanBuildFrom
+import scala.language.higherKinds
+import scala.util.{Failure, Success, Try}
+import scalaz.Scalaz._
+import scalaz.{Failure => FailureZ, Success => SuccessZ, _}
+
+case class DockerHash(hashType: String, hashString: String) {
+  val digest = s"$hashType:$hashString"
+}
+
+object DockerHash {
+  /**
+    * Creates a unique hash from a sequence of hashes.
+    * If there is a single hash, returns just the single hash string.
+    * If there is more than one hash, concatenates the hash strings together, and then md5s the resulting string.
+    * Returns a failure if there are zero hashes, or if there are more than one type of hash type.
+    *
+    * @param hashCollectionType A description of the collection.
+    * @param dockerHashes The sequence of hashes to rehash.
+    * @return A single docker hash.
+    */
+  def fromSeq(hashCollectionType: String, dockerHashes: Seq[DockerHash]): Try[DockerHash] = {
+    if (dockerHashes.isEmpty) {
+      // Need at least on docker hash to hash.
+      Failure(new IllegalArgumentException("docker hashes is empty"))
+    } else if (dockerHashes.size == 1) {
+      // Just use the original hash, but proceed it with the collection type.
+      val dockerHash = dockerHashes.head
+      Success(dockerHash.copy(s"$hashCollectionType-${dockerHash.hashType}"))
+    } else {
+      // Concatenate the hash strings together, then md5 the result to create another hash.
+      val hashTypes = dockerHashes.map(_.hashType).distinct
+      if (hashTypes.size == 1) {
+        val hashType = s"$hashCollectionType-${hashTypes.head}-md5"
+        val hashString = dockerHashes.map(_.hashString).mkString("").md5Sum
+        Success(DockerHash(hashType, hashString))
+      } else {
+        Failure(new IllegalArgumentException(s"found more than one docker hash type: $hashTypes"))
+      }
+    }
+  }
+
+  /**
+    * Creates a unique hash from a sequence of hashes.
+    * If there is a single hash, returns just the single hash string.
+    * If there is more than one hash, concatenates the hash strings together, and then md5s the resulting string.
+    * Returns a failure if there are zero hashes, or if there are more than one type of hash type.
+    *
+    * @param hashCollectionType A description of the collection.
+    * @param dockerHashes The sequence of hashes to rehash.
+    * @return A single docker hash.
+    */
+  def fromTries(hashCollectionType: String, dockerHashes: Seq[Try[DockerHash]]): Try[DockerHash] = {
+    trySequence(dockerHashes) flatMap {
+      fromSeq(hashCollectionType, _)
+    }
+  }
+
+  /** Copied from Future.sequence */
+  private def trySequence[A, M[X] <: TraversableOnce[X]](in: M[Try[A]])
+                                                (implicit cbf: CanBuildFrom[M[Try[A]], A, M[A]]): Try[M[A]] = {
+    in.foldLeft(Try(cbf(in))) {
+      (tr, ta) => for (r <- tr; a <- ta) yield r += a
+    } map (_.result())
+  }
```

Comment:
Thanks for the pointer! Switched over to using `TryUtil`. Will leave the version that uses a `mutable.Builder` for another day.